File: backend/app/models/complaint.py
import enum
import uuid
from datetime import datetime, timezone  # Include timezone

from sqlalchemy import Column, String, Float, DateTime, Enum, Text, Integer
from sqlalchemy.dialects.postgresql import UUID

from app._importer import load
Base = load("app.database.connection", "Base")


class Severity(str, enum.Enum):
    low = "Low"
    medium = "Medium"
    high = "High"


class Status(str, enum.Enum):
    reported = "Reported"
    verified = "Verified"
    assigned = "Assigned"
    repair_in_progress = "Repair in Progress"
    resolved = "Resolved"


from datetime import datetime, timezone  # Include timezone
# ... other imports ...

class Complaint(Base):
    __tablename__ = "complaints"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    image_url = Column(String, nullable=False)
    latitude = Column(Float, nullable=False)  # type: ignore
    longitude = Column(Float, nullable=False)  # type: ignore
    place_name = Column(Text, nullable=True)
    description = Column(Text, nullable=True)

    # AI output
    severity = Column(Enum(Severity), nullable=True)
    ai_confidence = Column(Float, nullable=True)  # type: ignore
    detections_count = Column(Integer, default=0)
    ai_summary = Column(Text, nullable=True)

    # Timestamps (replace .utcnow with timezone.utc to clear deprecation warnings)
    created_at = Column(DateTime, default=lambda: datetime.now(timezone.utc))
    updated_at = Column(DateTime, default=lambda: datetime.now(timezone.utc), onupdate=lambda: datetime.now(timezone.utc))

    # Workflow ───────────────────────────────────────────────────────────────
    status = Column(Enum(Status), default=Status.reported)
    assigned_department = Column(String, default="Unassigned")
    notes = Column(Text, nullable=True)

    # Duplicate clustering ───────────────────────────────────────────────────
    # Reports within DUPLICATE_RADIUS_METERS are grouped by the same cluster_id
    cluster_id = Column(UUID(as_uuid=True), nullable=True)

    # Push notifications ─────────────────────────────────────────────────────
    reporter_device_token = Column(String, nullable=True)

    

File: frontend/lib/models/complaint.dart
class Complaint {
  final String id;
  final String imageUrl;
  final double latitude;
  final double longitude;
  final String? placeName;
  final String? description;

  // AI output
  final String? severity;
  final double? aiConfidence;
  final int detectionsCount;
  final String? aiSummary;

  // Workflow
  final String status;
  final String assignedDepartment;
  final String? notes;

  final DateTime createdAt;
  final DateTime updatedAt;

  Complaint({
    required this.id,
    required this.imageUrl,
    required this.latitude,
    required this.longitude,
    this.placeName,
    this.description,
    this.severity,
    this.aiConfidence,
    required this.detectionsCount,
    this.aiSummary,
    required this.status,
    required this.assignedDepartment,
    this.notes,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Complaint.fromJson(Map<String, dynamic> json) => Complaint(
        id: json['id'] as String,
        imageUrl: json['image_url'] as String,
        latitude: (json['latitude'] as num).toDouble(),
        longitude: (json['longitude'] as num).toDouble(),
        placeName: json['place_name'] as String?,
        description: json['description'] as String?,
        severity: json['severity'] as String?,
        aiConfidence: json['ai_confidence'] == null
            ? null
            : (json['ai_confidence'] as num).toDouble(),
        detectionsCount: (json['detections_count'] as num?)?.toInt() ?? 0,
        aiSummary: json['ai_summary'] as String?,
        status: json['status'] as String? ?? 'Reported',
        assignedDepartment: json['assigned_department'] as String? ?? 'Unassigned',
        notes: json['notes'] as String?,
        createdAt: DateTime.parse(json['created_at'] as String),
        updatedAt: DateTime.parse(json['updated_at'] as String),
      );

  Complaint copyWith({
    String? status,
    String? assignedDepartment,
    String? notes,
  }) =>
      Complaint(
        id: id,
        imageUrl: imageUrl,
        latitude: latitude,
        longitude: longitude,
        placeName: placeName,
        description: description,
        severity: severity,
        aiConfidence: aiConfidence,
        detectionsCount: detectionsCount,
        aiSummary: aiSummary,
        status: status ?? this.status,
        assignedDepartment: assignedDepartment ?? this.assignedDepartment,
        notes: notes ?? this.notes,
        createdAt: createdAt,
        updatedAt: updatedAt,
      );
}